add cli parser documentation (#71497)
ansible/ansible: docs/docsite/rst/network/user_guide/cli_parsing.rst

.. _cli_parsing:

*****************************************

> Parsing Semi-structured Text with Ansible

Reviewer: ```suggestion
Parsing semi-structured text with Ansible
```

*****************************************

> The `ansible.netcommon <https://galaxy.ansible.com/ansible/netcommon>`_ collection version 1.2.0 or later  includes the :ref:`cli_parse <ansible_collections.ansible.netcommon.cli_parse_module>` module that can run commands and parse the output. You can use the ``cli_parse`` module on a device, host, or platform that only supports a command-line interface and the commands issued return semi-structured text.

Reviewer: ```suggestion
The `ansible.netcommon <https://galaxy.ansible.com/ansible/netcommon>`_ collection version 1.2.0 or later  includes the :ref:`cli_parse <ansible_collections.ansible.netcommon.cli_parse_module>` module that can run commands and parse the output. You can use the ``cli_parse`` module on a device, host, or platform that only supports a command-line interface. The commands you issue using this module return semi-structured text.
```

.. contents::
   :local:

Understanding the CLI parser
=============================

> The ``cli_parse`` module can either run a CLI command on a device and return a parsed result or can simply parse any text document. The ``cli_parse`` module includes cli_parser plugins to interface with a variety of parsing engines.

Reviewer: ```suggestion
The ``cli_parse`` module can run a CLI command on a device and return a parsed result. The module can also parse any text document without running any commands. The ``cli_parse`` module includes cli_parser plugins to interface with a variety of parsing engines.
```

> Why parse the text?

Reviewer: I think I might put this paragraph (the one below) at the top of the page as an intro, above the local TOC. It tells people why they want this feature.

Author: put part of it up there instead of the current text to keep the before toc text short but helpful  (and improve the google description, which should pull that first sentence or two in search results)

--------------------

> Parsing non-structured data such as network configurations into structured data allows programmatic use of the data from that device. You can pull information from a network device and update a CMDB in one playbook. Use cases include automated troubleshooting, creating dynamic documentation, updating IPAM (IP address management) tools and so on. You may prefer to do this with Ansible natively to take advantage of native Ansible constructs such as:

Reviewer: The page title refers to "semi-structured" text, but here we talk about transforming "non-structured" data into "structured" data. Are these two different use cases? If not, I'd suggest picking a term and using it in both places.

Author: fixed

-  The ``when`` clause to conditionally run other tasks or roles

> -  The ``assert`` module for checking configuration and operational state compliance

Reviewer: Parallelism:
```suggestion
-  The ``assert`` module to check configuration and operational state compliance
```

- The ``cli_parse`` module, by default, looks for the template in the templates directory as ``{{ short_os }}_{{ command }}.yaml``.

  - The ``short_os`` in the template filename is derived from either the host ``ansible_network_os`` or ``ansible_distribution``.
  - Spaces in the network or host command are replace with ``_`` in the ``command`` portion of the template filename. In this example, the ``show interfaces`` network CLI command becomes ``show_interfaces`` in the filename.

.. note::

	``ansible.netcommon.native`` parsing engine is fully supported with a Red Hat Ansible Automation Platform subscription.

Lastly in this task, the ``set_fact`` option sets the following ``interfaces`` fact for the device based on the now-structured data returned from ``cli_parse``:

.. code-block:: yaml

   Ethernet1/1:
       hardware: 100/1000/10000 Ethernet
       mac_address: 5254.005a.f8bd
       name: Ethernet1/1
       state:
       admin: up
       operating: up
   Ethernet1/10:
       hardware: 100/1000/10000 Ethernet
       mac_address: 5254.005a.f8c6
   <...>

Linux example
^^^^^^^^^^^^^

You can also use the native parser to run commands and parse output from Linux

> hosts.

Reviewer: There are a bunch of places like this where short paragraphs are broken into multiple lines. This will make translation more difficult. Can we consolidate a bit?

Author: the joy of automated md to rst conversion. I think I caught the remaining ones now

   - example: 'inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0'
     getval: |
      (?x)                                                 # free-spacing
      \s+inet\s(?P<inet>([0-9]{1,3}\.){3}[0-9]{1,3})       # the ip address
      /(?P<bits>\d{1,2})                                   # the mask bits
     result:
       "{{ name }}":
           ip_address: "{{ inet }}"
           mask_bits: "{{ bits }}"

.. note::

	 The ``shared`` key in the parser template allows the interface name to be used in subsequent parser entries. The use of examples and free-spacing mode with the regular expressions makes the template easier to read.

The following example task uses ``cli_parse`` with the native parser and the example template above to parse the Linux output:

.. code-block:: yaml

   - name: Run command and parse
     ansible.netcommon.cli_parse:
       command: ip addr show
       parser:
         name: ansible.netcommon.native
       set_fact: interfaces

This task assumes you previously gathered facts to determine the ``ansible_distribution`` needed to locate the template. Alternately, you could provide the path in the  ``parser/template_path`` option.

.. note::

> 	The ``ansible.netcommon.native`` parsing engine is fully supported with a Red Hat Ansible Automation Platform subscription.

Reviewer: Move information about support higher up, or maybe repeat in the list of available parsing engines? It feels a bit buried here and in the sections below about the other parsers. As a customer, I might choose only to read the detailed info for supported parsers, so making it easy to identify them is useful.

.. code-block:: yaml

   - name: "Run command and parse with ttp"
     ansible.netcommon.cli_parse:
       command: show interface
       parser:
         name: ansible.netcommon.ttp
       set_fact: interfaces

Taking a deeper dive in this task:

- The default template path ``templates/nxos_show_interface.ttp`` was generated using the ``ansible_network_os`` for the host and ``command`` provided.
- TTP supports several additional variables that will be passed to the parser. These include:

  - ``parser/vars/ttp_init`` -  Additional parameter passed when the parser is initialized.
  - ``parser/vars/ttp_results`` -  Additional parameters used to influence the parser output.
  - ``parser/vars/ttp_vars`` -  Additional variables made available in the template.

- See the `TTP documentation <https://ttp.readthedocs.io>`_ for details.

The task sets the follow fact as the ``interfaces`` fact for the
host:

.. code-block:: yaml

   - admin_state: up,
     interface: mgmt0
     state: up
   - admin_state: up,
     interface: Ethernet1/1
     state: up
   - admin_state: up,
     interface: Ethernet1/2
     state: up

Converting XML
-----------------

> Although Ansible contains a number of plugins that can convert XML to Ansible native data structures, the``cli_parse`` module runs the command on devices that return XML and returns the converted data in a single task.

Reviewer: This is a great comparison - I might mention this higher up in the doc as well, since consolidating two tasks into one is likely a big selling point for this approach.
